easybuild/tools/singularity/utilities.py

```diff
@@ -0,0 +1,391 @@
+    bootstrap_type = bootstrap_list[0]
+
+    # check bootstrap type value and ensure it is localimage, shub, docker
+    if bootstrap_type != "localimage" and bootstrap_type != "shub" and bootstrap_type != "docker":
+    	print " bootstrap type must be localimage, shub, or docker "
+	sys.exit(1)
+
+    # extracting application name,version, version suffix, toolchain name, toolchain version from
+    # easyconfig class
+
+    appname = ordered_ecs[0]['ec']['name']
+    appver = ordered_ecs[0]['ec']['version']
+    appversuffix = ordered_ecs[0]['ec']['versionsuffix']
+
+    tcname = ordered_ecs[0]['ec']['toolchain']['name']
+    tcver = ordered_ecs[0]['ec']['toolchain']['version']
+
+    osdeps = ordered_ecs[0]['ec']['osdependencies']
+
+    modulepath = ""
+
+
+    # with localimage it only takes 2 arguments. --singularity-bootstrap localimage:/path/to/image
+    # checking if path to image is valid and verify image extension is".img or .simg"
+    if bootstrap_type == "localimage":
+    	bootstrap_imagepath = bootstrap_list[1]
+	if os.path.exists(bootstrap_imagepath):
+		# get the extension of container image
+		image_ext = os.path.splitext(bootstrap_imagepath)[1]
+		if image_ext == ".img" or image_ext == ".simg":
+    			_log.debug("Image Extension is OK")
+		else:
+			print "Invalid image extension %s, must be .img or .simg", image_ext
+			raise EaasyBuildError("Invalid image extension %s must be .img or .simg", image_ext)
+	else:
+		
+		print "Can't find image path ", bootstrap_imagepath
+		raise EasyBuildError("Can't find image path %s", bootstrap_imagepath)
+
+    # if option is shub or docker		
+    else:
+	bootstrap_image = bootstrap_list[1]
+        image_tag = "NONE"
+    	# format --singularity-bootstrap shub:<image>:<tag>
+        if len(bootstrap_list) == 3:
+		image_tag = bootstrap_list[2]
+
+    module_scheme = get_module_naming_scheme()
+    
+    # bootstrap from local image
+    if bootstrap_type == "localimage":
+	bootstrap_content = "Bootstrap: " + bootstrap_type + " \n"
+	bootstrap_content += "From: " + bootstrap_imagepath + "\n" 
+    # default bootstrap is shub or docker
+    else:
+	    bootstrap_content = "BootStrap: " + bootstrap_type + "\n" 
+
+	    if image_tag == "NONE":
+		    bootstrap_content += "From: " + bootstrap_image  + "\n"
+	    else:
+		    bootstrap_content += "From: " + bootstrap_image + ":" + image_tag  + "\n"
+    
+    if module_scheme == "HierarchicalMNS":
+	    modulepath = "/app/modules/all/Core"
+    else:
+	    modulepath = "/app/modules/all/"
+
+    post_content = """
+%post
+"""
+    # if there is osdependencies in easyconfig then add them to Singularity recipe	
+    if len(osdeps) > 0:
+    	# format: osdependencies = ['libibverbs-dev', 'libibverbs-devel', 'rdma-core-devel']
+        if isinstance(osdeps[0],basestring):
+	     	for os_package in osdeps:
+		     	post_content += "yum install -y " + os_package + " || true \n"
```

Comment:
You're assuming that's ok for one of the dependencies installation will fail. Is that correct? And if so why?

Reply:
exactly, I want singularity to build the container without failing on yum install since osdependencies from easyconfig has options to install os package based on distro Debian vs Fedora. Keep in mind this feature is not perfect, I am not sure if eb --robot with osdependencies are picking all os packages from all easyconfigs. Right now I am able to get os packages from the easyconfig. I need this for installing OpenMPI